### PR TITLE
Input Validation for Rules

### LINKED ISSUE #104: Input Validation for Rule Parser
As a PyReason user writing rules for logic programs, I want the rule parser to validate all inputs thoroughly and provide clear, actionable error messages, so that I can catch syntax errors, malformed rules, and invalid parameters early in development rather than encountering cryptic runtime errors or silent failures during reasoning.                                                                                   
                                                                                               
  Acceptance Criteria                                                                          
                                                                                               
  Core Rule Syntax Validation                                                                  
                                                                                               
  - Rule text must be a valid, non-empty string with proper arrow separator syntax             
  - Head and body components must be present and properly formatted                            
  - Predicates and variables must follow valid naming conventions                              
  - Parentheses and delimiters must be properly matched and positioned                         
                                                                                               
  Body Clause Validation                                                                       
                                                                                               
  - Body clauses must be well-formed with no empty or malformed entries                        
  - Clause separators (commas) must be used correctly                                          
  - Comparison operators, when present, must be properly structured                            
                                                                                               
  Bound and Interval Validation                                                                
                                                                                               
  - Interval bounds must contain valid numeric values within acceptable ranges                 
  - Bound format must follow the expected syntax                                               
  - Special values (NaN, Inf, negative numbers) must be handled appropriately                  
                                                                                               
  Optional Parameter Validation                                                                
                                                                                               
  - Custom thresholds (when provided) must be valid and match clause count                     
  - Weights array (when provided) must be numeric, properly formatted, and match clause count  
  - Special quantifiers (e.g., forall) must follow correct syntax                              
                                                                                               
  Error Handling                                                                               
                                                                                               
  - All validation errors produce clear, actionable error messages                             
  - Error messages specify what went wrong and hint at correct usage                           
  - Parser fails fast on first validation error                                                
                                                                                               
  Testing & Quality                                                                            
                                                                                               
  - Comprehensive unit test coverage for valid and invalid inputs                              
  - Edge cases and boundary conditions tested                                                  
  - Code passes linting and style checks                                                       
                                                                                               
  Definition of Done                                                                           
                                                                                               
  - All acceptance criteria met                                                                
  - Unit tests passing with good coverage                                                      
  - Linting passing                                                                            
  - Code reviewed and approved

### PR DESCRIPTION
**Pull request by ColtonPayne**
## Summary

Closes #104

Adds comprehensive input validation to the rule parser (`rule_parser.py`) so that malformed rules, invalid bounds, and bad parameters produce clear, actionable error messages instead of cryptic runtime failures.

### Changes

- **Entry-point validation** — rule text must be a non-empty string with exactly one `<-` separator; head and body must both be non-empty after splitting
- **Predicate & variable name validation** — names must match `[a-zA-Z_][a-zA-Z0-9_]*`; digit-leading names and special characters are rejected with specific messages
- **Bound validation** — intervals must contain exactly two finite numeric values in `[0, 1]` with `lower <= upper`; NaN, Inf, and out-of-range values are caught
- **Negation with explicit bounds** — `~pred(X):[l,u]` now correctly computes `[1-u, 1-l]` for both head and body clauses
- **`:True` / `:False` shorthand** — body and head bounds accept boolean string literals as aliases for `[1,1]` and `[0,0]`
- **Double negation detection** — `~~` in head or body raises immediately
- **`forall()` syntax validation** — the inner expression must contain a predicate with variables
- **Threshold validation** — empty dicts, negative keys, and length mismatches raise `ValueError`
- **Weight validation** — must be numeric, finite, non-negative, and match clause count; plain lists are auto-converted to `np.float64` arrays
- **Error type cleanup** — generic `Exception` raises replaced with `TypeError` / `ValueError`
- **Refactored internals** — extracted `_split_body_into_clauses`, `_parse_head`, `_validate_predicate_name`, and `_validate_variable_name` helpers
- **Makefile typo fix** — corrected `don_disable_jit` → `dont_disable_jit` in test suite target

### Test coverage

Added 722-line test file (`tests/unit/dont_disable_jit/test_rule_parser.py`) with three test classes:
- **`TestValidRuleParsing`** — node/edge rules, temporal deltas, annotation functions, bounds, negation, `:True`/`:False`, `forall`, head functions, thresholds, weights, and a complex combined-parameter case
- **`TestInvalidRuleParsing`** — type errors, empty inputs, missing arrows, bad bounds, wrong-length thresholds/weights, NaN/Inf weights, malformed `forall`
- **`TestEdgeCasesAndBoundary`** — boundary bounds, large delta_t, all six comparison operators, single-char predicates, negation inversion arithmetic, NaN/Inf bounds, digit-leading names, double negation

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: pyreason/scripts/utils/rule_parser.py [hunk 1]
```diff
@@ -8,135 +10,125 @@
...
-        # Add body variables depending on whether there's an operator or not
-        variables = clause[start_idx+1:end_idx].split(',')
-        start_idx = clause.find('(', start_idx+1)
-        end_idx = clause.find(')', end_idx+1)
+    for clause_str in body_clauses:
+        # V8: Body clause must contain parentheses
+        start_idx = clause_str.find('(')
+        if start_idx == -1:
+            raise ValueError(f"Body clause '{clause_str}' must contain parentheses around variables")
+
```
Comment (kmukherji) [inline pyreason/scripts/utils/rule_parser.py]: Similar to previous comment, we should be careful when using the term variable. It has a specific meaning in logic.
